lib/wallet/src/Cardano/Wallet/Write/Tx/Redeemers.hs

```diff
@@ -0,0 +1,237 @@
+            indexedRedeemers <- StateT assignNullRedeemers
+            executionUnits <- get
+                >>= lift . evaluateExecutionUnits indexedRedeemers
+            modifyM (assignExecutionUnits executionUnits)
+            modify' addScriptIntegrityHash
+        pure $ Cardano.ShelleyTx shelleyBasedEra ledgerTx'
+  where
+    epochInformation :: EpochInfo (Either T.Text)
+    epochInformation =
+        hoistEpochInfo (left (T.pack . show)) $ epochInfo timeTranslation
+
+    systemStart = systemStartTime timeTranslation
+
+    -- | Assign redeemers with null execution units to the input transaction.
+    --
+    -- Redeemers are determined from the context given to the caller via the
+    -- 'Redeemer' type which is mapped to an 'Alonzo.ScriptPurpose'.
+    assignNullRedeemers
+        :: RecentEraLedgerConstraints (ShelleyLedgerEra era)
+        => Tx (ShelleyLedgerEra era)
+        -> Either ErrAssignRedeemers
+            ( Map Alonzo.RdmrPtr Redeemer
+            , Tx (ShelleyLedgerEra era)
+            )
+    assignNullRedeemers ledgerTx = do
+        (indexedRedeemers, nullRedeemers) <-
+            fmap unzip $ forM redeemers parseRedeemer
+        pure
+            ( Map.fromList indexedRedeemers
+            , ledgerTx
+                & witsTxL . rdmrsTxWitsL
+                    .~ (Alonzo.Redeemers (Map.fromList nullRedeemers))
+            )
+      where
+        parseRedeemer rd = do
+            let mPtr = Alonzo.rdptr
+                    (txBody (recentEra @era) ledgerTx)
+                    (toScriptPurpose rd)
+            ptr <- case mPtr of
+                SNothing -> Left $ ErrAssignRedeemersTargetNotFound rd
+                SJust ptr -> pure ptr
+            let mDeserialisedData =
+                    deserialiseOrFail $ BL.fromStrict $ redeemerData rd
+            rData <- case mDeserialisedData of
+                Left e -> Left $ ErrAssignRedeemersInvalidData rd (show e)
+                Right d -> pure (Alonzo.Data d)
+            pure ((ptr, rd), (ptr, (rData, mempty)))
+
+    -- | Evaluate execution units of each script/redeemer in the transaction.
+    -- This may fail for each script.
+    evaluateExecutionUnits
+        :: RecentEraLedgerConstraints (ShelleyLedgerEra era)
+        => Map Alonzo.RdmrPtr Redeemer
+        -> Tx (Cardano.ShelleyLedgerEra era)
+        -> Either ErrAssignRedeemers
+            (Map Alonzo.RdmrPtr (Either ErrAssignRedeemers Alonzo.ExUnits))
+    evaluateExecutionUnits indexedRedeemers ledgerTx = do
+        let res =
+                Ledger.evalTxExUnits
+                    pparams
+                    ledgerTx
+                    (fromCardanoUTxO utxo)
+                    epochInformation
+                    systemStart
+        case res of
+            Left translationError ->
+                Left $ ErrAssignRedeemersTranslationError translationError
+            Right report ->
+                Right $ hoistScriptFailure indexedRedeemers report
```

Comment:
I know this PR just moves the functions, but why not:

```suggestion
    evaluateExecutionUnits indexedRedeemers ledgerTx =
        Ledger.evalTxExUnits
            pparams ledgerTx (fromCardanoUTxO utxo) epochInformation systemStart
        & bimap
            ErrAssignRedeemersTranslationError
            (hoistScriptFailure indexedRedeemers)
```

Reply:
🤔 fair enough 👍